feat(vertical-spacing): General approach for content spacing - OEL-1955

tibi2303 opened the pull request:
In this pr:
- Management of vertical spacing ( all content inside pages are inside a div with class .bcl-section that take cares of the spacing between texts and headings and also components - banners, listings )
- Common root namespace for all the components that use headings in order for the default spacing between heading and content

tibi2303 commented on src/themes/default/src/scss/_section.scss:
```diff
@@ -0,0 +1,37 @@
+  > h2,
+  > .bcl-heading {
+    margin-bottom: $bcl-spacing-m;
+    margin-top: $bcl-spacing-m;
```
<img width="885" alt="image" src="https://user-images.githubusercontent.com/36895476/185875917-51b174ca-b073-416b-98bb-f1ac6416a93a.png">
For the container class names to remain clean, we can't go for a mono directional approach.

tibi2303 replied:
Most i can do is to have this margin-top only on the .bcl-section:first-child, in order to get the spacing.

planctus commented on src/themes/default/src/scss/_section.scss:
```diff
@@ -0,0 +1,59 @@
+    margin-top: $bcl-spacing-l;
+  }
+
+  > .bcl-listing,
```
this is the section component css, why would we style bcl-banner and bcl-listing..? can't we wrap those "components" into a section to get the desiderd spacing..?

planctus commented on src/components/bcl-accordion/accordion.html.twig:
```diff
@@ -5,7 +5,7 @@
   - title_tag (string) (default: 'h2')
   - title_link: (link object) (default: {})
-  - title_attributes: (drupal attrs)
+  - title_attributes: (drupal attrs) (default: classes: "mb-4")
```
it seems we are still confused about these headings, i'll try to explain the way i see this:
We have an include of the bcl-heading in all the paragraphs templates. 
**These headings should get their style from the section css, .bcl-section h(n), generic enough to trigger all the headings regardless of the fact that they are direct children of the section wrapper**.
This class that you add here then it would override this default style, this would only be justified in case we have a specific requirement, which we don't have, so i expect these utilities to be removed, basically.

Then there might be other headings as part of the component markup, which is the case here for the accordion, if we do like i propose this will then require to reset the default margins on all the other headings, this would be handled adding a utility class to the internal headings (the buttons).

Furthermore in the stories where we demo the paragraphs, including the title, in the story file the mb-4 should be added in order to ensure some spacing for the sake of the demo.

planctus commented on src/compositions/bcl-glossary/glossary-detail.html.twig:
```diff
@@ -28,16 +28,15 @@
 	} only %}
 
 	{% for file in files %}
 		{% if loop.last %}
```
we should try to avoid this, potentially we could use the utility we added, for them it seems difficult to add a class to all the items in the array but the last one, so we have to approach this differently

planctus commented on src/compositions/bcl-group/group-landing.html.twig:
```diff
@@ -74,13 +74,19 @@
-      <h2 class="mt-4 mt-md-0 mb-4 pb-2">{{ members.title}}</h2>
+      {% include  '@oe-bcl/bcl-heading/heading.html.twig' with {
+        title: members.title,
+        title_tag: 'h2',
```
All these h2(s) can be removed, it's the default in the heading template.

planctus commented on src/compositions/bcl-group/package.json:
```diff
@@ -16,6 +16,7 @@
     "@openeuropa/bcl-listing": "^0.26.0",
     "@openeuropa/bcl-navigation": "^0.26.0",
     "@openeuropa/bcl-recent-activities": "^0.26.0",
+    "@openeuropa/bcl-heading": "^0.26.0",
```
not so important but can we keep an alphabetical order..?

planctus commented on src/compositions/bcl-event/event.html.twig:
```diff
@@ -5,20 +5,28 @@
+      title: 'Description',
+      attributes: create_attribute().setAttribute('id', 'item-1'),
+    } only %}
+    <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Morbi sit amet eleifend tortor. In facilisis eros vitae turpis ullamcorper,
```
utility for generating random text has been added also to twig as a function "get_dummy_text()" with the same params as the js utility.

planctus commented on src/compositions/bcl-glossary/glossary-detail.html.twig:
```diff
@@ -1,43 +1,39 @@
-	Fusce commodo hendrerit diam, pretium tempus leo varius sit amet. Etiam interdum, orci at sagittis luctus, 
-	lorem libero tempus mauris, a fermentum libero orci semper lacus. Duis tristique fringilla magna,
-	eu egestas dolor molestie non. </p>
+		<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse
```
also here, you can use the utility

planctus commented on src/themes/default/src/scss/_section.scss:
```diff
@@ -0,0 +1,64 @@
+    margin-bottom: 0;
+  }
+
+  &:first-child {
```
i don't think this makes really sense..i don't understand the logic and indeed the demo is broken, showing a doubled margin. If we are in a margin-bottom only approach, then the upper section is going to create the spacing for the one below, regardless of the fact that they are siblings. This that you did would only if the sections are all siblings and then margin-collapse would take care of "eatring" the doubled spacing. Applying this in the "real world", the content banner would be wrapped in a seciton, this will create the spacing with the left sidebar and the main content, no need to have an internal spacing in the section first item, then...

tibi2303 replied:
indeed, i removed it and added the banner inside a section and it seems ok

planctus commented on src/themes/default/src/scss/_section.scss:
```diff
@@ -0,0 +1,60 @@
+    &:first-child {
+      margin-top: 0;
+    }
+    &:last-child {
```
This breaks, i mean, it would make sense only if the heading is the last element of the section, but like this we are actually resetting the margin-bottom of the last h2, most likely this is wrong, considering that there is "content" below it.

tibi2303 replied:
i think it may indeed happen if the content is not wrapped, in any case i removed it.

planctus replied:
You only changed this for the minor headings, but the same applies to the h1, h2 headings, the margin-bottom should not be reset for the last-child and first-of-type should be used instead of first-child

planctus commented on src/themes/default/src/scss/_section.scss:
```diff
@@ -0,0 +1,57 @@
+  .bcl-heading {
+    margin-bottom: $bcl-spacing-m;
+    margin-top: $bcl-spacing-m;
+    &:first-child {
```
i think that what you want to have here is :first-of-type more than first:child, you want to reset the margin of the first heading, first-child is only going to target the heading when it is the first element in the section...

planctus commented on src/compositions/bcl-contact-form/contact-form.html.twig:
```diff
@@ -1,5 +1,7 @@
 
 {% block content %}
-  {% include '@oe-bcl/bcl-form/form.html.twig' with form only %}
+  <div class="bcl-section">
```
let's use <section>, instead of a generic div

planctus commented on src/themes/default/src/scss/_section.scss:
```diff
@@ -0,0 +1,54 @@
+  }
+}
+
+@include media-breakpoint-down(md) {
```
we have a problem in mobile, using the section in the left sidebar is creating an empty area due to the margins. Then i don't understand the values you used, it is just a demo, but why did you set a bigger margin in mobile than we have in desktop..? If we want to demo this "feature", we would probably do the opposite, no..?

tibi2303 replied:
I lowered the spacing to 1rem.

planctus commented on src/themes/default/src/scss/_section.scss:
```diff
@@ -0,0 +1,54 @@
+  @include clearfix;
+  margin-bottom: $bcl-spacing-l;
+
+  h1,
```
the fact that you put also h1 here is adding margin to the "page title" in the content banner, which is not necessarily bad, but then we use the my in the "subtitle" and now it is not needed anymore, can be simply mb-

tibi2303 replied:
All of the muted text under the titles should have now mb-

tibi2303 replied:
still i am thinking now that the spacing is not equal

planctus commented on src/themes/default/src/scss/_section.scss:
```diff
@@ -0,0 +1,54 @@
+  .bcl-section {
+    margin-bottom: $bcl-spacing-xl;
+
+    &:first-of-type {
```
but then why are we adding this margin here..?

tibi2303 replied:
Indeed, by removing this, it should also fix that big spacing from the sidebar

tibi2303 replied:
removed

tibi2303 commented:
I modified the margins of the footer to align them with figma on compositions.
I changed the file aproach - an extra margin class was added from a bad drupal attributes class adition.
